File: grobots_cpp/views/GBScoresView.cpp
// GBSideView.cpp
// Grobots (c) 2002-2004 Devon and Warren Schudy
// Distributed under the GNU General Public License.

#include "GBScoresView.h"
#include "GBSide.h"
#include "GBWorld.h"
#include "GBRobotType.h"
#include "GBStringUtilities.h"

const short kGraphHeight = 120;
const short kInfoBoxHeight = 184;
const short kTotalsHeight = 56;

void GBScoresView::DrawIncome(const GBIncomeStatistics & income,
		short left, short right, short top) {
	long total = income.Total();
	if ( ! total ) return;
	DrawStringPair("Solar:", ToPercentString(income.Autotrophy(), total),
		left, right, top + 10, 9, GBColor::darkGreen);
	DrawStringPair("Manna:", ToPercentString(income.Theotrophy(), total),
		left, right, top + 20, 9, GBColor::darkGreen);
	DrawStringPair("Enemies:", ToPercentString(income.Heterotrophy(), total),
		left, right, top + 30, 9, GBColor::purple);
	DrawStringPair("Stolen:", ToPercentString(income.Kleptotrophy(), total),
		left, right, top + 40, 9, GBColor(0.4f, 0.6f, 0));
	DrawStringPair("Cannibal:", ToPercentString(income.Cannibalism(), total),
		left, right, top + 50, 9, GBColor::darkRed);
}

void GBScoresView::DrawExpenditures(const GBExpenditureStatistics & spent,
		short left, short right, short top) {
	long total = spent.Total();
	if ( ! total ) return;
	DrawStringPair("Growth:", ToPercentString(spent.Construction(), total),
		left, right, top + 10, 9, GBColor::darkGreen);
	DrawStringPair("Engine:", ToPercentString(spent.Engine(), total),
		left, right, top + 20, 9, GBColor::black);
	DrawStringPair("Sensors:", ToPercentString(spent.Sensors(), total),
		left, right, top + 30, 9, GBColor::blue);
	DrawStringPair("Weapons:", ToPercentString(spent.Weapons(), total),
		left, right, top + 40, 9, GBColor::purple);
	DrawStringPair("Force:", ToPercentString(spent.ForceField(), total),
		left, right, top + 50, 9, GBColor::blue);
	DrawStringPair("Shield:", ToPercentString(spent.Shield(), total),
		left, right, top + 60, 9, GBColor::blue);
	DrawStringPair("Repairs:", ToPercentString(spent.Repairs(), total),
		left, right, top + 70, 9, GBColor::black);
	DrawStringPair("Brains:", ToPercentString(spent.Brain(), total),
		left, right, top + 80, 9, GBColor::black);
	DrawStringPair("Stolen:", ToPercentString(spent.Stolen(), total),
		left, right, top + 90, 9, GBColor(0.4f, 0.6f, 0));
	DrawStringPair("Overflow:", ToPercentString(spent.Wasted(), total),
		left, right, top + 100, 9, GBColor::darkRed);
}

void GBScoresView::DrawDeaths(const GBScores & scores,
		short left, short right, short top) {
	DrawStringPair("Kills:", ToPercentString(scores.KilledFraction()),
		left, right, top + 10, 9, GBColor::purple);
	if ( scores.Survived() )
		DrawStringPair("Relative:", ToPercentString(scores.KillRate(), 0),
			left, right, top + 20, 9, GBColor::black);
	DrawStringLongPair("Kills:", scores.Killed(), left, right, top + 30, 9, GBColor::purple);
	DrawStringLongPair("Dead:", scores.Dead(), left, right, top + 40, 9, GBColor::black);
	DrawStringLongPair("Suicide:", scores.Suicide(), left, right, top + 50, 9, GBColor::darkRed);
}

void GBScoresView::DrawGraph(const GBRect & box, long vscale, int hscale,
							 const std::vector<long> & hist, const GBColor & color) {
	int n = hist.size() - 1;
//draw lines
	for ( int i = 0; i < n; ++ i )
		DrawLine(box.left + box.Width() * i / hscale,
			box.bottom - hist[i] * box.Height() / vscale,
			box.left + box.Width() * (i + 1) / hscale,
			box.bottom - hist[i + 1] * box.Height() / vscale,
			color);
}

void GBScoresView::DrawGraphs(const GBRect & box) {
	if ( ! world.Sides() ) return;
	DrawBox(box);
	GBRect graph = box;
	graph.Shrink(1);
	const GBSide * side = world.SelectedSide();
	long scale = 1;
	int hscale = 1;
	if ( side ) {
		const std::vector<long> hist = side->Scores().BiomassHistory();
		for ( int i = 0; i < hist.size(); ++i )
			if (hist[i] > scale)
				scale = hist[i];
		hscale = hist.size();
	//average biomass, if available
		int rounds = side->TournamentScores().Rounds();
		if ( rounds ) {
			const std::vector<long> avg = side->TournamentScores().BiomassHistory();
			if ( avg.size() > hscale )
				hscale = avg.size();
			for ( int i = 0; i < avg.size(); ++i )
				if ( avg[i] > scale )
					scale = avg[i];
			DrawGraph(graph, scale, hscale - 1, avg,
				rounds > 20 ? GBColor::black : GBColor::darkGray);
		}
		DrawGraph(graph, scale, hscale - 1, hist, side->Color().ContrastingTextColor());
	} else { //all sides
		bool allRounds = graphAllRounds && world.TournamentScores().Rounds();
		const GBSide * s;
		for ( s = world.Sides(); s; s = s->next ) {
			if ( (allRounds ? s->TournamentScores().Rounds() : s->Scores().Rounds()) == 0 )
				continue;
			const std::vector<long> hist = (allRounds ? s->TournamentScores().BiomassHistory() : s->Scores().BiomassHistory());
			for ( int i = 0; i < hist.size(); ++i )
				if (hist[i] > scale)
					scale = hist[i];
			if ( hist.size() > hscale )
				hscale = hist.size();
		}
		for ( s = world.Sides(); s; s = s->next ) {
			if ( (allRounds ? s->TournamentScores().Rounds() : s->Scores().Rounds()) == 0 )
				continue;
			DrawGraph(graph, scale, hscale - 1,
				(allRounds ? s->TournamentScores().BiomassHistory() : s->Scores().BiomassHistory()),
				s->Color().ContrastingTextColor());
		}
	}
	DrawStringLeft(ToString(scale), box.left + 4, box.top + 13, 10, GBColor::darkGray);
}

void GBScoresView::DrawRoundScores(const GBScores & scores, const GBRect & box) {
	short c1 = box.left + 3;
	short c2 = (box.left + box.right) / 2 + 2;
// basics
	DrawStringLongPair("Biomass:", scores.Biomass(), c1, c2 - 4, box.top + 25, 9, GBColor::darkGreen);
	DrawStringLongPair("Population:", scores.Population(), c1, c2 - 4, box.top + 35, 9, GBColor::blue);
	DrawStringLongPair("Ever:", scores.PopulationEver(), c1 + 10, c2 - 4, box.top + 45, 9, GBColor::blue);
	const GBSide * side = world.SelectedSide();
	if ( side ) {
		if ( side->Scores().Sterile() && side->Scores().SterileTime() != side->Scores().ExtinctTime() )
			DrawStringLongPair("Sterile:", side->Scores().SterileTime(),
				c1, c2 - 4, box.top + 60, 9, GBColor::purple);
		if ( ! side->Scores().Population() )
			DrawStringLongPair("Extinct:", side->Scores().ExtinctTime(),
				c1, c2 - 4, box.top + 70, 9, GBColor::red);
	}
// income
	DrawStringLongPair("Income:", scores.Income().Total(), c1, c2 - 4, box.top + 95, 9, GBColor::black);
	DrawIncome(scores.Income(), c1, c2 - 4, box.top + 95);
	DrawStringLongPair("Seed:", scores.Seeded(), c1, c2 - 4, box.top + 155, 9, GBColor::black);
	if ( scores.Efficiency() > 0 )
		DrawStringPair("Efficiency:", ToPercentString(scores.Efficiency(), 0),
			c1, c2 - 4, box.top + 170, 9, GBColor::black);
	if ( scores.Doubletime(world.CurrentFrame()) )
		DrawStringLongPair("Double:", scores.Doubletime(world.CurrentFrame()),
			c1, c2 - 4, box.top + 180, 9, GBColor::black);
// expenditures
	DrawStringLongPair("Spent:", scores.Expenditure().Total(), c2, box.right - 3, box.top + 25, 9, GBColor::black);
	DrawExpenditures(scores.Expenditure(), c2, box.right - 3, box.top + 25);
// death
	DrawDeaths(scores, c2, box.right - 3, box.top + 130);
}

void GBScoresView::DrawTournamentScores(const GBScores & tscores, const GBRect & box) {
	short c3 = box.left + 3;
	short c4 = (box.left + box.right) / 2 + 2;
	DrawStringLongPair("Rounds:", tscores.Rounds(), c3, c4 - 4, box.top + 10, 9, GBColor::black);
	if (tscores.Rounds()) {
		DrawStringPair("Biomass:", ToPercentString(tscores.BiomassFraction()),
			c3, c4 - 4, box.top + 25, 9, GBColor::darkGreen);
		DrawStringPair("Early:", ToPercentString(tscores.EarlyBiomassFraction()),
			c3 + 10, c4 - 4, box.top + 35, 9, GBColor::darkGreen);
		DrawStringPair("Survival:", ToPercentString(tscores.Survival()),
			c3, c4 - 4, box.top + 50, 9, GBColor::black);
		DrawStringPair("Early death:", ToPercentString(tscores.EarlyDeathRate()),
			c3, c4 - 4, box.top + 60, 9, GBColor::black);
		DrawStringPair("Late death:", ToPercentString(tscores.LateDeathRate()),
			c3, c4 - 4, box.top + 70, 9, GBColor::black);
	// income
		DrawStringLongPair("Avg income:", tscores.Income().Total() / tscores.Rounds(),
			c3, c4 - 4, box.top + 95, 9, GBColor::darkGreen);
		DrawIncome(tscores.Income(), c3, c4 - 4, box.top + 95);
		DrawStringLongPair("Avg seed:", tscores.Seeded(),
			c3, c4 - 4, box.top + 155, 9, GBColor::black);
		if ( tscores.Efficiency() > 0 )
			DrawStringPair("Efficiency:", ToPercentString(tscores.Efficiency(), 0),
				c3, c4 - 4, box.top + 170, 9, GBColor::black);
	// expenditures
		DrawStringLongPair("Avg spent:", tscores.Expenditure().Total() / tscores.Rounds(),
			c4, box.right - 4, box.top + 25, 9, GBColor::black);
		DrawExpenditures(tscores.Expenditure(), c4, box.right - 4, box.top + 25);
	// death
		DrawDeaths(tscores, c4, box.right - 4, box.top + 130);
	}
}

void GBScoresView::Draw() {
	const GBSide * side = world.SelectedSide();
	DrawBackground();
	GBRect graphbox(kEdgeSpace, kEdgeSpace, Width() - kEdgeSpace, kGraphHeight + kEdgeSpace);
//round statistics
	GBRect box(kEdgeSpace, graphbox.bottom + kEdgeSpace,
		(Width() - kEdgeSpace) / 2, graphbox.bottom + kEdgeSpace + kInfoBoxHeight);
	DrawBox(box);
	DrawStringLeft(side ? side->Name() : std::string("Overall statistics"),
		box.left + 3, box.top + 13, 12, GBColor::black);
// draw stats...
	if ( side )
		DrawRoundScores(side->Scores(), box);
	else
		DrawRoundScores(world.RoundScores(), box);
// tournament stats:
	box.left = box.right + kEdgeSpace;
	box.right = Width() - kEdgeSpace;
	DrawBox(box);
	if ( side )
		DrawTournamentScores(side->TournamentScores(), box);
	else
		DrawTournamentScores(world.TournamentScores(), box);
// simulation total values
	box.top = box.bottom + kEdgeSpace;
	box.bottom = box.top + kTotalsHeight;
	DrawBox(box);
	DrawStringLeft("Total values:", box.left + 3, box.top + 13, 10);
	DrawStringLongPair("Manna:", world.MannaValue(),
		box.left + 3, box.right - 3, box.top + 23, 10, GBColor::darkGreen);
	DrawStringLongPair("Corpses:", world.CorpseValue(),
		box.left + 3, box.right - 3, box.top + 33, 10, GBColor::red);
	DrawStringLongPair("Robots:", world.RobotValue(),
		box.left + 3, box.right - 3, box.top + 43, 10, GBColor::blue);
// object counts
	box.left = kEdgeSpace;
	box.right = (Width() - kEdgeSpace) / 2;
	DrawBox(box);
	DrawStringLongPair("Robots:", world.CountObjects(ocRobot),
		box.left + 3, box.right - 3, box.top + 13, 10);
	DrawStringLongPair("Foods:", world.CountObjects(ocFood),
		box.left + 3, box.right - 3, box.top + 23, 10);
	DrawStringLongPair("Shots:", world.CountObjects(ocShot) + world.CountObjects(ocArea),
		box.left + 3, box.right - 3, box.top + 33, 10);
	DrawStringLongPair("Sensors:", world.CountObjects(ocSensorShot),
		box.left + 3, box.right - 3, box.top + 43, 10);
	DrawStringLongPair("Decorations:", world.CountObjects(ocDecoration),
		box.left + 3, box.right - 3, box.top + 53, 10);
//drawing graph last to reduce flicker
	DrawGraphs(graphbox);
// record
	lastDrawnWorld = world.ChangeCount();
	lastSideDrawn = world.SelectedSide();
	graphAllLastDrawn = graphAllRounds;
}

GBScoresView::GBScoresView(GBWorld & w)
	: world(w),
	lastDrawnWorld(-1), lastSideDrawn(nil),
	graphAllLastDrawn(true), graphAllRounds(true)
{}

GBMilliseconds GBScoresView::RedrawInterval() const {
	return 2000;
}

bool GBScoresView::InstantChanges() const {
	return lastSideDrawn != world.SelectedSide() || graphAllRounds != graphAllLastDrawn;
}

bool GBScoresView::DelayedChanges() const {
	return lastDrawnWorld != world.ChangeCount();
}

short GBScoresView::PreferredWidth() const {
	return 400;
}

const string GBScoresView::Name() const {
	return "Statistics";
}

short GBScoresView::PreferredHeight() const {
	return kGraphHeight + kInfoBoxHeight + kTotalsHeight + kEdgeSpace * 4;
}

File: src/views/GBScoresView.java
package views;

/*******************************************************************************
 * Copyright (c) 2002-2013 (c) Devon and Warren Schudy Copyright (c) 2014
 * Devon and Warren Schudy, Mike Anderson
 *******************************************************************************/
import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.Rectangle;
import java.util.List;

import javax.swing.BorderFactory;
import javax.swing.GroupLayout;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextPane;
import javax.swing.border.Border;
import javax.swing.text.BadLocationException;
import javax.swing.text.Document;
import javax.swing.text.SimpleAttributeSet;
import javax.swing.text.Style;
import javax.swing.text.StyleConstants;

import sides.GBScores;
import sides.Side;
import simulation.GBGame;
import support.GBColor;
import support.StringUtilities;
import ui.SideSelectionListener;

public class GBScoresView extends JPanel implements SideSelectionListener {
	/**
	 * 
	 */
	private static final long serialVersionUID = 4827315972785490558L;

	GBGame game;
	Side selectedSide;

	JLabel roundLabel, tournLabel;
	JPanel roundGraph, tournGraph;
	JTextPane roundIncome, roundKills, roundExpenditure, roundCounts;
	JTextPane tournIncome, tournKills, tournExpenditure, tournCounts;

	Style basicStyle;
	SimpleAttributeSet basicAttr;
	SimpleAttributeSet bold;
	SimpleAttributeSet blue;
	SimpleAttributeSet green;
	SimpleAttributeSet red;
	SimpleAttributeSet gray;
	SimpleAttributeSet brown;

	public GBScoresView(GBGame _game) {
		game = _game;
		Border border = BorderFactory.createLineBorder(getForeground(), 1);
		roundLabel = new JLabel();
		tournLabel = new JLabel();
		roundGraph = new JPanel() {
			/**
			 * 
			 */
			private static final long serialVersionUID = -3268061604878534231L;
			Point max = new Point(0, 0);
			int rounds = 0;

			@Override
			public void paintComponent(Graphics g) {
				super.paintComponent(g);
				if (game.getTournamentScores().rounds > rounds)
					max = new Point(0, 0);
				drawGraph((Graphics2D) g, getVisibleRect(), false, max);
			};
		};
		roundGraph.setBorder(border);
		roundGraph.setPreferredSize(new Dimension(200, 150));
		roundGraph.setBackground(Color.white);
		roundGraph.setIgnoreRepaint(true);
		tournGraph = new JPanel() {
			/**
			 * 
			 */
			private static final long serialVersionUID = 3274144226270291895L;
			Point max = new Point(0, 0);

			@Override
			public void paintComponent(Graphics g) {
				super.paintComponent(g);
				drawGraph((Graphics2D) g, getVisibleRect(), true, max);
			};
		};
		tournGraph.setBorder(border);
		tournGraph.setPreferredSize(new Dimension(200, 150));
		tournGraph.setBackground(Color.white);
		tournGraph.setIgnoreRepaint(true);
		roundIncome = new JTextPane();
		roundIncome.setBorder(border);
		roundKills = new JTextPane();
		roundKills.setBorder(border);
		roundExpenditure = new JTextPane();
		roundExpenditure.setBorder(border);
		roundCounts = new JTextPane();
		roundCounts.setBorder(border);
		tournIncome = new JTextPane();
		tournIncome.setBorder(border);
		tournKills = new JTextPane();
		tournKills.setBorder(border);
		tournExpenditure = new JTextPane();
		tournExpenditure.setBorder(border);
		tournCounts = new JTextPane();
		tournCounts.setBorder(border);
		GroupLayout layout = new GroupLayout(this);
		setLayout(layout);
		layout.setAutoCreateGaps(true);
		layout.setAutoCreateContainerGaps(true);
		layout.setHorizontalGroup(layout
				.createSequentialGroup()
				.addGroup(
						layout.createParallelGroup()
								.addComponent(roundLabel)
								.addComponent(roundGraph)
								.addGroup(
										layout.createSequentialGroup()
												.addGroup(
														layout.createParallelGroup()
																.addComponent(
																		roundIncome)
																.addComponent(
																		roundKills))
												.addComponent(roundExpenditure)
												.addComponent(roundCounts)))
				.addGroup(
						layout.createParallelGroup()
								.addComponent(tournLabel)
								.addComponent(tournGraph)
								.addGroup(
										layout.createSequentialGroup()
												.addGroup(
														layout.createParallelGroup()
																.addComponent(
																		tournIncome)
																.addComponent(
																		tournKills))
												.addComponent(tournExpenditure)
												.addComponent(tournCounts))));
		layout.setVerticalGroup(layout
				.createParallelGroup()
				.addGroup(
						layout.createSequentialGroup()
								.addComponent(roundLabel)
								.addComponent(roundGraph)
								.addGroup(
										layout.createParallelGroup()
												.addGroup(
														layout.createSequentialGroup()
																.addComponent(
																		roundIncome)
																.addComponent(
																		roundKills))
												.addComponent(roundExpenditure)
												.addComponent(roundCounts)))
				.addGroup(
						layout.createSequentialGroup()
								.addComponent(tournLabel)
								.addComponent(tournGraph)
								.addGroup(
										layout.createParallelGroup()
												.addGroup(
														layout.createSequentialGroup()
																.addComponent(
																		tournIncome)
																.addComponent(
																		tournKills))
												.addComponent(tournExpenditure)
												.addComponent(tournCounts))));
		setAttributes();
		setIgnoreRepaint(true);
	}

	void setAttributes() {
		basicAttr = new SimpleAttributeSet();
		StyleConstants.setFontSize(basicAttr, 9);
		bold = new SimpleAttributeSet(basicAttr);
		StyleConstants.setBold(bold, true);
		red = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(red, Color.red);
		green = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(green, GBColor.darkGreen);
		blue = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(blue, Color.blue);
		red = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(red, Color.red);
		gray = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(gray, Color.gray);
		brown = new SimpleAttributeSet(basicAttr);
		StyleConstants.setForeground(brown, new Color(150, 113, 23));
	}

	public void updateScores() {
		roundLabel.setText(String.format("%s This Round:",
				selectedSide == null ? "" : selectedSide.getName()));
		tournLabel.setText(String.format("%s Average over %d rounds:",
				selectedSide == null ? "" : selectedSide.getName(),
				game.getTournamentScores().rounds));
		GBScores scores = selectedSide == null ? game.getRoundScores()
				: selectedSide.getScores();
		drawIncome(roundIncome, scores);
		drawKills(roundKills, scores);
		drawExpenditure(roundExpenditure, scores);
		drawCounts(roundCounts, scores);
		scores = selectedSide == null ? game.getTournamentScores() : selectedSide
				.getTournamentScores();
		drawIncome(tournIncome, scores);
		drawKills(tournKills, scores);
		drawExpenditure(tournExpenditure, scores);
		drawCounts(tournCounts, scores);
	}

	void drawIncome(JTextPane pane, GBScores score) {
		pane.setText("");
		Document doc = pane.getDocument();
		try {
			double total = score.income.total() != 0 ? score.income.total() / 100.0
					: 1;
			doc.insertString(doc.getLength(), "Income\t", bold);
			doc.insertString(doc.getLength(),
					String.format("%d\n", score.income.total()), bold);
			doc.insertString(doc.getLength(), "Solar\t", green);
			doc.insertString(doc.getLength(),
					String.format("%.1f%%\n", score.income.autotrophy / total),
					green);
			doc.insertString(doc.getLength(), "Manna\t", green);
			doc.insertString(doc.getLength(),
					String.format("%.1f%%\n", score.income.theotrophy / total),
					green);
			doc.insertString(doc.getLength(), "Enemies\t", basicAttr);
			doc.insertString(doc.getLength(), String.format("%.1f%%\n",
					score.income.heterotrophy / total), basicAttr);
			doc.insertString(doc.getLength(), "Stolen\t", green);
			doc.insertString(doc.getLength(), String.format("%.1f%%\n",
					score.income.kleptotrophy / total), green);
			doc.insertString(doc.getLength(), "Cannibal\t", red);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.income.cannibalism / total),
					red);
		} catch (BadLocationException e) {
			e.printStackTrace();
		}
	}

	void drawKills(JTextPane pane, GBScores score) {
		pane.setText("");
		Document doc = pane.getDocument();
		try {
			doc.insertString(doc.getLength(), "Kill Rate\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.0f%%\n", score.getKilledFraction()),
					basicAttr);
			if (score.survived != 0) {
				doc.insertString(doc.getLength(), "Relative\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.0f%%\n", score.getKillRate()), basicAttr);
			} else
				doc.insertString(doc.getLength(), "\n", basicAttr);
			doc.insertString(doc.getLength(), "Kills\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.0f\n", score.killed), basicAttr);
			doc.insertString(doc.getLength(), "Losses\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.0f\n", score.dead), basicAttr);
			doc.insertString(doc.getLength(), "Suicides\t", red);
			doc.insertString(doc.getLength(),
					String.format("%.0f\n", score.suicide), red);
		} catch (BadLocationException e) {
			e.printStackTrace();
		}
	}

	void drawExpenditure(JTextPane pane, GBScores score) {
		pane.setText("");
		Document doc = pane.getDocument();
		try {
			double total = score.expenditure.total() != 0 ? score.income
					.total() / 100.0 : 1;
			doc.insertString(doc.getLength(), "Spent\t", bold);
			doc.insertString(doc.getLength(),
					String.format("%d\n", score.expenditure.total()), bold);
			doc.insertString(doc.getLength(), "Growth\t", green);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.construction
							/ total), green);
			doc.insertString(doc.getLength(), "Engine\t", basicAttr);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.engine / total),
					basicAttr);
			doc.insertString(doc.getLength(), "Sensors\t", blue);
			doc.insertString(doc.getLength(), String.format("%.1f%%\n",
					score.expenditure.sensors / total), blue);
			doc.insertString(doc.getLength(), "Weapons\t", basicAttr);
			doc.insertString(doc.getLength(), String.format("%.1f%%\n",
					score.expenditure.weapons / total), basicAttr);
			doc.insertString(doc.getLength(), "Force\t", blue);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.forceField
							/ total), blue);
			doc.insertString(doc.getLength(), "Shield\t", basicAttr);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.shield / total),
					basicAttr);
			doc.insertString(doc.getLength(), "Repairs\t", basicAttr);
			doc.insertString(doc.getLength(), String.format("%.1f%%\n",
					score.expenditure.repairs / total), basicAttr);
			doc.insertString(doc.getLength(), "Brains\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.brain / total),
					basicAttr);
			doc.insertString(doc.getLength(), "Stolen\t", brown);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.stolen / total),
					brown);
			doc.insertString(doc.getLength(), "Overflow\t", red);
			doc.insertString(
					doc.getLength(),
					String.format("%.1f%%\n", score.expenditure.wasted / total),
					red);
		} catch (BadLocationException e) {
			e.printStackTrace();
		}
	}

	void drawCounts(JTextPane pane, GBScores score) {
		pane.setText("");
		Document doc = pane.getDocument();
		try {
			if (selectedSide == null) {
				doc.insertString(doc.getLength(), "Biomass\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.0f\n", score.biomass), basicAttr);
				doc.insertString(doc.getLength(), "Population\t", blue);
				doc.insertString(doc.getLength(),
						String.format("%d\n", score.population), blue);
				doc.insertString(doc.getLength(), "Ever\t", blue);
				doc.insertString(doc.getLength(),
						String.format("%d\n", score.getPopulationEver()), blue);
				doc.insertString(doc.getLength(), "Manna\t", green);
				doc.insertString(doc.getLength(),
						String.format("%d\n", game.getMannaValue()), green);
				doc.insertString(doc.getLength(), "Corpses\t", red);
				doc.insertString(doc.getLength(),
						String.format("%d\n", game.getCorpseValue()), red);
			} else {
				doc.insertString(doc.getLength(), "Biomass\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.1f%%\n", score.biomassFraction * 100),
						basicAttr);
				doc.insertString(doc.getLength(), "Early\t", blue);
				doc.insertString(doc.getLength(),
						String.format("%.0f\n", score.earlyBiomass), blue);
				doc.insertString(doc.getLength(), "Survival\t", blue);
				doc.insertString(doc.getLength(),
						String.format("%.1f%%\n", score.getSurvival() * 100), blue);
				doc.insertString(doc.getLength(), "Early Death\t", green);
				doc.insertString(
						doc.getLength(),
						String.format("%.1f%%\n", score.EarlyDeathRate() * 100),
						green);
				doc.insertString(doc.getLength(), "Late Death\t", red);
				doc.insertString(doc.getLength(),
						String.format("%.1f%%\n", score.getLateDeathRate() * 100),
						red);
			}
			doc.insertString(doc.getLength(), "Seeded\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.0f\n", score.seeded), basicAttr);
			doc.insertString(doc.getLength(), "Efficiency\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%.1f%%\n", score.getEfficiency() * 100),
					basicAttr);
			int doubletime = score.getDoubletime(game.currentFrame());
			if (selectedSide != null && doubletime != 0 && doubletime < 100000) {
				doc.insertString(doc.getLength(), "DoubleTime\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.d\n", doubletime), basicAttr);
			} else
				doc.insertString(doc.getLength(), "\n", basicAttr);
			if (score.population != 0) {
				doc.insertString(doc.getLength(), "Economy\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.1f%%\n", score.getEconFraction()),
						basicAttr);
				doc.insertString(doc.getLength(), "Combat\t", basicAttr);
				doc.insertString(doc.getLength(),
						String.format("%.1f%%\n", score.getCombatFraction()),
						basicAttr);
			} else
				doc.insertString(doc.getLength(), "\n\n", basicAttr);
			doc.insertString(doc.getLength(), "Territory\t", basicAttr);
			doc.insertString(doc.getLength(),
					String.format("%d\n", score.territory), basicAttr);
		} catch (BadLocationException e) {
			e.printStackTrace();
		}
	}

	@Override
	public void paintComponent(Graphics g) {
		super.paintComponent(g);
		roundGraph.repaint();
		tournGraph.repaint();
	}

	void drawGraph(Graphics2D g, Rectangle rect, boolean allRounds, Point max) {
		if (game.getSidesCount() == 0)
			return;
		rect.grow(-10, -5);
		Side side = selectedSide;
		// Check for changes to the max sizes
		if (allRounds)
			max.y = 0;
		for (Side s : game.sides) {
			if ((allRounds ? s.getTournamentScores().rounds : s.getScores().rounds) == 0)
				continue;
			List<Integer> hist = allRounds ? s.getTournamentScores()
					.getBiomassHistory() : s.getScores().getBiomassHistory();
			if (hist.size() == 0)
				continue;
			if (hist.size() - 1 > max.x) {
				max.x = hist.size() - 1;
			}
			if (!allRounds) {
				max.y = Math.max(hist.get(hist.size() - 1), max.y);
			} else {
				for (Integer i : hist)
					max.y = Math.max(max.y, i);
			}
		}
		if (max.x <= 0) {
			// max.x = 0;
			max.y = 0;
			return;
		}
		// draw gridlines
		// Vertical
		for (int t = 50; t < max.x; t += 50) {
			int x = (int) (rect.width * (double) t / max.x);
			g.setColor(Color.lightGray);
			g.drawLine(x, rect.y, x, rect.height);
		}
		// Horizontal
		for (int quantum = 1000; quantum < max.y; quantum *= 10) {
			if (quantum < max.y / 40)
				continue;
			for (int en = quantum; en < max.y; en += quantum) {
				int y = rect.height - (int) (rect.height * (double) en / max.y);
				g.setColor(Color.lightGray);
				g.drawLine(rect.x, y, rect.x + rect.width, y);
			}
		}
		// draw curves
		for (Side s : game.sides) {
			if ((allRounds ? s.getTournamentScores().rounds : s.getScores().rounds) == 0)
				continue;
			List<Integer> hist = (allRounds ? s.getTournamentScores()
					.getBiomassHistory() : s.getScores().getBiomassHistory());
			// draw lines

			for (int j = 0; j < hist.size() - 1; ++j) {
				g.setColor(GBColor.ContrastingTextColor(s.getColor()));
				g.setStroke(new BasicStroke(s == side ? 2 : 1));
				g.drawLine(
						(int) (rect.width * (double) j / max.x),
						rect.height
								- (int) ((double) rect.height
										* (double) hist.get(j) / max.y),
						(int) (rect.width * (double) (j + 1) / max.x),
						rect.height
								- (int) ((double) rect.height
										* (double) hist.get(j + 1) / max.y));
			}
		}
		StringUtilities.drawStringLeft(g, Integer.toString(max.y), 3, 10, 9,
				Color.gray);
		StringUtilities.drawStringRight(g, Integer.toString(max.x * 100),
				(int) rect.getWidth() - 3, (int) rect.getHeight() - 3, 9,
				Color.gray);
	}

	@Override
	public void setSelectedSide(Side side) {
		selectedSide = side;
		updateScores();
		if (isVisible())
			repaint();
	}
}
